### DISCUSSION TITLE
Extend Relation Controller delete query

### DISCUSSION THREAD

**Discussion by ericp-mrel** [wintercms/winter#719] [Q&A] [answered]
Hi,

Does anyone know if it is possible to override or extend the relation manager delete action? I have some records which are seeded as part of the system and should be read-only, but the user is able to add their own custom records which they should be able to delete. I use a `is_system_managed` column in the DB to denote which ones are editable or not editable.

I tried to use controller middleware, but it seems it runs after the relation controller runs this method `onRelationManageDelete` which seems strange to me.

Curious if anyone knows of a way to prevent users from deleting certain records via the relation manager widget. Or other solutions that could work?

Thanks!

**Comment by mjauvin** [comment]
What about using this to filter-out those records?

https://wintercms.com/docs/backend/relations#extend-filter-widgets

Then use this on the widget object:
```
$widget->bindEvent('list.extendQuery', function ($query) {
    ...
});
```

**Comment by mjauvin** [comment]
The RelationController also has this code:
```
            if ($sqlConditions = $this->getConfig('manage[conditions]')) {
                $widget->bindEvent('list.extendQueryBefore', function ($query) use ($sqlConditions) {
                    $query->whereRaw($sqlConditions);
                }); 
            }
            elseif ($scopeMethod = $this->getConfig('manage[scope]')) {
                $widget->bindEvent('list.extendQueryBefore', function ($query) use ($scopeMethod) {
                    $query->$scopeMethod($this->model);
                }); 
            }   
```

So you should be able to use the `conditions` or `scope` relation config settings to modify the query.

ref. https://wintercms.com/docs/backend/relations#configuring-relation

**Reply by ericp-mrel** [reply]
The problem with doing something like this, is I want those records to still show up in the list because users can add/remove them from the relation. I just don't want them to be able to delete the system managed ones, and I still want them to be able to delete the user created ones.

Basically, I want to populate common options by default and allow users to create their own custom ones, but only allow them to delete the options they've created.

Hope that makes sense.

**Reply by ericp-mrel** [reply]
I think I may have found a work-around which works for my use case. While it probably is pretty hacky, it at least works since the `relationExtendManageWidget()` controller method is run before the `RelationController` runs the `onRelationManageDelete()` method.

```php
# BackendController.php

// ...

public function relationExtendManageWidget($widget, $field, $model)
{
    // Do not allow backend users to delete system managed Plan Amenities.
    if ($this->getAjaxHandler() === 'onRelationButtonDelete' && $field === 'foo') {
            $ids = Foo::query()
                ->where('is_system_managed', false)
                ->whereIn('id', post('checked'))
                ->lists('id');

            request()->request->set('checked', $ids);
    }
}
```

I just have a question because I'm curious. Is it normal / expected for the [controller middleware](https://wintercms.com/docs/backend/controllers-ajax#controller-middleware) to be run after the Relation controller has already deleted the records? To me this seems like a bug since the middleware should run first so you can filter or do whatever to the request before any ajax actions are actually performed such as deleting records.

**Reply by mjauvin** [reply]
Then I would use the `backend.ajax.beforeRunHandler` event. The handler should be `onRelationButtonDelete` I believe

ref. https://wintercms.com/docs/events/event/backend.ajax.beforeRunHandler

**Comment by mjauvin** [comment]
Can you show how you tried to use the middleware in this case?

**Reply by ericp-mrel** [reply]
Here's how I tried to use the middleware:

```php
class TestController extends Controller
{
    public $implement = [
        'Backend\Behaviors\ListController',
        'Backend\Behaviors\FormController',
        'Backend\Behaviors\RelationController',
        'Backend\Behaviors\ImportExportController',
    ];

    /** @var string|array */
    public $listConfig = 'config_list.yaml';

    /** @var string|array */
    public $formConfig = 'config_form.yaml';

    /** @var string|array */
    public $relationConfig = 'config_relation.yaml';

    /** @var string|array */
    public $importExportConfig = 'config_import_export.yaml';

    public function __construct()
    {
        parent::__construct();

        $this->middleware(function (Request $request, Response $response) {
            trace_log('controller middleware');

            dd(get_class($request), $response);
        });
    }
}
```

I added some logging statements inside the RelationController and that middlware function

```php
// RelationController.php

    /**
     * Delete an existing related model completely
     */
    public function onRelationManageDelete()
    {
        $this->beforeAjax();

        trace_log('onRelationManageDelete');

        /*
         * Multiple (has many, belongs to many)
         */
        if ($this->viewMode === 'multi') {
            if (($checkedIds = post('checked')) && is_array($checkedIds)) {
                foreach ($checkedIds as $relationId) {
                    if (!$obj = $this->relationModel->find($relationId)) {
                        continue;
                    }

                    trace_log('$obj->delete()');
                    // $obj->delete();
                }
            }
        }
        /*
         * Single (belongs to, has one)
         */
        elseif ($this->viewMode === 'single') {
            $relatedModel = $this->viewModel;
            if ($relatedModel->exists) {
                trace_log('$relatedModel->delete()');
                // $relatedModel->delete();
            }

            // Reinitialise the form with a blank model
            $this->initRelation($this->model);

            $this->viewWidget->setFormValues([]);
            $this->viewModel = $this->relationModel;
        }

        return $this->relationRefresh();
    }
```

and here's the log output
```
[2022-09-28 20:08:00] local.INFO: onRelationManageDelete  
[2022-09-28 20:08:00] local.INFO: $obj->delete()  
[2022-09-28 20:08:00] local.INFO: $obj->delete()  
[2022-09-28 20:08:00] local.INFO: $obj->delete()  
[2022-09-28 20:08:00] local.INFO: $obj->delete()  
[2022-09-28 20:08:00] local.INFO: $obj->delete()  
[2022-09-28 20:08:00] local.INFO: $obj->delete()  
[2022-09-28 20:08:00] local.INFO: $obj->delete()  
[2022-09-28 20:08:00] local.INFO: $obj->delete()  
[2022-09-28 20:08:00] local.INFO: controller middleware  
```

This shows the middleware being executed after the records have been deleted.

**Comment by mjauvin** [comment]
The documentation says this regarding middleware:
```
Controller middleware is executed after the request is processed by Winter CMS, but before the response is sent to the browser.
```
ref. https://wintercms.com/docs/backend/controllers-ajax#controller-middleware

**Comment by ericp-mrel** [comment, marked as answer]
This works for my needs: https://github.com/wintercms/winter/discussions/719#discussioncomment-3756014

> Then I would use the backend.ajax.beforeRunHandler event. The handler should be onRelationButtonDelete I believe
>
> ref. https://wintercms.com/docs/events/event/backend.ajax.beforeRunHandler

Thanks for the suggestion @mjauvin
